MM-10160 - support interactive button styles

RyanCommits opened the pull request:
#### Summary
Supports the customization of interactive message buttons through the `style` property on `attachments.actions` objects. Dummy test actions used to achieve screenshot:
```
[ 
   { 
      id:'gpc1tihowbfmmcizofm7zhr1to',
      type:'button',
      name:'dog',
      style:'danger'
   },
   { 
      id:'gpc1tihowddto',
      type:'button',
      name:'cat',
      style:'primary'
   },
   { 
      id:'gpc1tihowbfmmcizddo',
      type:'button',
      name:'moose'
   },
   { 
      id:'gpc1tihowbfawemmcizddo',
      type:'button',
      name:'chocolate',
      style:'#166de0'
   },
   { 
      id:'gpc1tihowfabfmmcizddo',
      type:'button',
      name:'rat',
      style:'onlineIndicator'
   }
]
```

**Things to note:**
1. I expanded the `STATUS_COLORS` object to define the "default" and "primary" colors, and picked the original default blue color. If you want those definitions changed, let me know.

2. I assume the server will send me a valid `style` property if it exists. If we don't trust the response, I can do some sort of validation.

3. Button default styles on the web app: 
![Screen Shot 2020-02-06 at 7 06 05 PM](https://user-images.githubusercontent.com/30220112/73989449-c9f98280-4913-11ea-978a-04614075caf0.png)

Notice on my own screenshot on button 3 "moose", the default style is a button with white text and blue background. If we want to change the default style to something like the web, something without background color, let me know.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/13260
https://mattermost.atlassian.net/browse/MM-10160

#### Checklist
- [x] Has UI changes

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) --> 
iPhone 11 Pro Max Simulator - 13.3

#### Screenshots
**After**:
![Screen Shot 2020-02-06 at 6 48 53 PM](https://user-images.githubusercontent.com/30220112/73989006-5e62e580-4912-11ea-98cf-97439e7f69b5.png)

**Before**:
![Screen Shot 2020-02-06 at 7 09 52 PM](https://user-images.githubusercontent.com/30220112/73989590-4ee49c00-4914-11ea-9672-52d82fb04d69.png)

enahum commented on app/components/message_attachments/message_attachment.js:
```diff
@@ -18,10 +18,12 @@ import AttachmentThumbnail from './attachment_thumbnail';
 import AttachmentFooter from './attachment_footer';
 
-const STATUS_COLORS = {
+export const STATUS_COLORS = {
```
lets move this constant to a file under the constants folder

RyanCommits replied:
done

matthewbirtch commented:
@enahum what did you want me to review with this? I don't have much context or background.

enahum commented:
@mattewbirtch the UI changes I guess (the buttons that reflect the options of the poll)... are we ok with them? Do they need anything else? It does not look very refined to me.

enahum commented:
Also wondering if there is a limit to the amount of options, if there isn’t should there be?

matthewbirtch commented:
@enahum can we open a UX ticket to look at this? I'd like to understand all the uses of these interactive buttons and how/why we're proposing custimization. I'd also like to look at the styles we're using here and provide a better default I think.

lieut-data commented:
@matthewbirtch, see https://mattermost.atlassian.net/browse/MM-10160 for context, but this is largely an effort to expand our offering to match Slack's behaviour here. Their guidelines encourage plugin authors to use colours sparingly, but afford them the flexibility to pick whatever colours they want.

enahum commented:
@RyanCommits until we figure out what to do with the UI, can you please make the requested changes on the code.
Thanks.

lieut-data commented:
@matthewbirtch, to confirm, each button can be styled with a separate colour. Any concerns applying  the thrust of your changes that show the buttons all having the same colour?

matthewbirtch commented:
I was hoping to avoid the ugliness of the rainbow effect, but if we don't want to impose any limitation here, it's fine. Here's how that looks
<table>
<tr>
<td width="50%"><img src="https://user-images.githubusercontent.com/2040554/74189966-70e56380-4c1f-11ea-9383-470ec5870b54.png"></td>
<td></td>
</tr>
</table>

matthewbirtch commented:
I also wanted to provide how this displays in the web app/desktop. This layout is also in the Figma file linked above.
![web-app-poll-all-colors](https://user-images.githubusercontent.com/2040554/74191841-5660b980-4c22-11ea-9e78-922fc739c0d4.png)

jasonblais commented:
cc @aaronrothschild @itao @esethna on above designs from product team

RyanCommits commented:
> @RyanCommits until we figure out what to do with the UI, can you please make the requested changes on the code.
> Thanks.

Unit tests added and the requested changes are completed. Code updated to match designs.

<img width="294" alt="Screen Shot 2020-02-16 at 3 12 18 PM" src="https://user-images.githubusercontent.com/30220112/74612001-1771ae80-50cf-11ea-87a6-b555c96a56d2.png">

matthewbirtch commented:
@RyanCommits @enahum is there a build I can test with?

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/3716  
https://circleci.com/gh/mattermost/mattermost-mobile/3715

enahum commented:
@matthewbirtch should be posted in the QA channel soon

mattermod commented:
Artifact links:  
https://3716-70265724-gh.circle-artifacts.com/0/%7E/mattermost-mobile/Mattermost_Beta.ipa  
https://3715-70265724-gh.circle-artifacts.com/0/%7E/mattermost-mobile/Mattermost_Beta.apk

RyanCommits commented:
@matthewbirtch Answer to your questions:
1. This feature on the webapp is not implemented. We'll need someone to do that.
2. Testing the various colors on a pre-built app installed in release mode is not possible unless mattermost already supports customizing button colors on the server side. Maybe @lieut-data can confirm. The way I'm testing this is locally hijacking the response from the server, and injecting my own data. I don't know of any other way to replicate the colors. I can give you an updated screen shot if it helps: 
![Screen Shot 2020-02-19 at 6 36 09 PM](https://user-images.githubusercontent.com/30220112/74887015-b43e8100-5347-11ea-9255-8eaf3879fd06.png)

**Padding tweaks**: done
**Hard coded hex values**: "good" and "warning" were properties that existed before. I didn't add them myself. I replaced the hardcoded values with constants from `mattermost-redux`

RyanCommits commented:
Here's the android screenshot: Pixel 2xl

If you look at the left padding of the buttons, you'll notice that it's a tiny bit bigger compared to the right. For some reason, the outer container's big blue border on the left side is pushing the entire content to the right. If I remove the blue border, the extra space goes away. 

Not sure why this is a thing, perhaps a bug with android.

![Screenshot_20200219-190320](https://user-images.githubusercontent.com/30220112/74888026-c8d04880-534a-11ea-9f6d-ad55b5a26627.png)

RyanCommits commented:
@matthewbirtch Thanks for the diligent review!

**Addressing your concerns**:
1. It seems like Android does not respect `fontWeight: 600`. I checked other elements that are `600` and they all show up thin. This may be a bigger issue that needs addressing. For example, "hey", which is the title of the poll, should be `fontWeight: 600`, but shows up thin on Android as well. This bug is in the production version of the app too.

2. Done. The buttons themselves were 38px rather than 36px as per the design. I updated them to be 36px in height.

**ios:**
![Screen Shot 2020-02-20 at 6 48 15 PM](https://user-images.githubusercontent.com/30220112/74990594-b240f480-5411-11ea-9ad1-cabb5534b180.png)

**android**
![Screenshot_20200220-184824](https://user-images.githubusercontent.com/30220112/74990581-ab19e680-5411-11ea-8320-ab3933726526.png)

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/3767  
https://circleci.com/gh/mattermost/mattermost-mobile/3766

mattermod commented:
Artifact links:  
https://3767-70265724-gh.circle-artifacts.com/0/%7E/mattermost-mobile/Mattermost_Beta.apk  
https://3766-70265724-gh.circle-artifacts.com/0/%7E/mattermost-mobile/Mattermost_Beta.ipa

chuttam commented on app/components/message_attachments/action_button/action_button.js:
```diff
@@ -20,6 +21,7 @@ export default class ActionButton extends PureComponent {
         theme: PropTypes.object.isRequired,
         cookie: PropTypes.string.isRequired,
         disabled: PropTypes.bool,
+        style: PropTypes.string,
```
`style` is a very overloaded word, and tends to be reserved for theming/styling of entire components as a whole, rather than just a singular colour styling element, as is the case here.

Could we rename to `actionColor`, `actionButtonColor` or something similar? Maybe even just `color` is appropriate.

RyanCommits replied:
Thank you! Renamed to `buttonColor`.

chuttam commented on app/components/message_attachments/action_button/action_button.js:
```diff
@@ -28,19 +30,27 @@ export default class ActionButton extends PureComponent {
 
     render() {
-        const {name, theme, disabled} = this.props;
+        const {name, theme, disabled, style: customStyle} = this.props;
```
If `style` is renamed, this simplifies and conflict with theme `style` is automatically handled.

RyanCommits commented on app/components/message_attachments/action_button/action_button.test.js:
```diff
@@ -0,0 +1,77 @@
+
+            const buttonTextChild = wrapper.getElement().props.children;
+
+            expect(wrapper.getElement().props.containerStyle[0].borderColor).toBe(changeOpacity(Preferences.THEMES.default.centerChannelColor, 0.25));
```
done!

RyanCommits commented on app/components/message_attachments/action_button/action_button.test.js:
```diff
@@ -0,0 +1,79 @@
+        },
+    ];
+
+    actions.forEach(({id, name, style}) => {
```
I see the value in that. In this format the code does not communicate explicitly that there are four possibilities to test: data without style, data with hex style, data with theme style, and data with status style.

I can expand this single test to four tests if it improves readability?

RyanCommits replied:
updated

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/3817  
https://circleci.com/gh/mattermost/mattermost-mobile/3816

mattermod commented:
Artifact links:  
https://3817-70265724-gh.circle-artifacts.com/0/%7E/mattermost-mobile/Mattermost_Beta.ipa  
https://3816-70265724-gh.circle-artifacts.com/0/%7E/mattermost-mobile/Mattermost_Beta.apk
